aten/src/ATen/core/boxing/impl/boxing.h

```diff
@@ -5,75 +5,247 @@
-    named_not_supported_kernel(functor, opHandle, nullptr);  // does not return
+struct BoxedKernelWrapper<Result(Args...), std::enable_if_t<has_dimname_arg<Args...>::value, void>> {
+  static Result call(KernelFunction::InternalBoxedKernelFunction*, OperatorKernel*, const OperatorHandle&, Args... args) {
+    TORCH_INTERNAL_ASSERT(false, "Call to a boxed kernel with unboxable parameter type at::Dimname.");
   }
+};
 
-  TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::call() for a kernel that only has a boxed kernel and doesn't support calling from an unboxed API yet.");
-}
+// at::Quantizer
+template <class... Args>
+using has_quantizer_arg =
+  guts::disjunction<
+    std::is_same<at::Quantizer, std::decay_t<Args>>...,
+    std::is_same<c10::intrusive_ptr<at::Quantizer>, std::decay_t<Args>>...
+  >;
 
-// SFINAE version for ops with returns
 template<class Result, class... Args>
-std::enable_if_t<supports_boxing<Result, Args...>::value && !std::is_same<void, Result>::value, Result>
-boxAndCallBoxedFunc(KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func, OperatorKernel* functor, const OperatorHandle& opHandle, Args... args) {
-  // TODO Reuse stack vector instead of allocating?
-  torch::jit::Stack stack;
-  torch::jit::push(stack, std::forward<Args>(args)...);
-
-  (*boxed_kernel_func)(functor, opHandle, &stack);
-
-  TORCH_INTERNAL_ASSERT(stack.size() == 1, "A boxed kernel should only push one return to the stack");
-  return std::move(stack[0]).to<Result>();
-}
+struct BoxedKernelWrapper<Result(Args...), std::enable_if_t<has_quantizer_arg<Args...>::value, void>> {
+  static Result call(KernelFunction::InternalBoxedKernelFunction*, OperatorKernel*, const OperatorHandle&, Args... args) {
+    TORCH_INTERNAL_ASSERT(false, "Unboxed call to a boxed kernel with unboxable parameter type at::Quantizer.");
+  }
+};
 
-// SFINAE version for ops without returns
+// 2. Supported signatures, other than ref-passing.
+//
 template<class Result, class... Args>
-std::enable_if_t<supports_boxing<Result, Args...>::value && std::is_same<void, Result>::value, Result>
-boxAndCallBoxedFunc(KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func, OperatorKernel* functor, const OperatorHandle& opHandle, Args... args) {
-  // TODO Reuse stack vector instead of allocating?
-  torch::jit::Stack stack;
-  torch::jit::push(stack, std::forward<Args>(args)...);
-
-  (*boxed_kernel_func)(functor, opHandle, &stack);
+struct BoxedKernelWrapper<
+  Result(Args...),
+  std::enable_if_t<
+    can_box_all<Args...>::value && can_unbox<Result>::value && !is_tuple_of_tensor_refs<Result>::value,
+    void
+  >
+> {
+  static Result call(
+    KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func,
+    OperatorKernel* functor,
+    const OperatorHandle& opHandle,
+    Args... args
+  ) {
+    // TODO Reuse stack vector instead of allocating?
+    torch::jit::Stack stack;
+    stack.reserve(sizeof...(Args));
```

Comment:
In some cases certain argument types convert into multiple stack entries (`TensorOptions` is the most notable example):

```
static inline void push_one(Stack& stack, c10::TensorOptions options) {
  stack.emplace_back(c10::typeMetaToScalarType(options.dtype()));
  stack.emplace_back(options.layout());
  stack.emplace_back(options.device());
  stack.emplace_back(options.pinned_memory());
}
```

In those cases, the reservation computed using `sizeof...` would be too small. Beware!

Reply:
😮 

Excellent point. We could notice TensorOptions in the args and over-reserve in that case (absent an exact member count), but I think we want to try to go to zero allocations rather than keeping it to just 1, so I'm less inclined to deal with the TO case until we try to move the whole thing to the machine stack, or TLS, or whatever. 

But definitely worth noting that stack size >= argc. cc @smessmer @ljk53 re IValue stack tactics